### ISSUE TITLE
`selector-max-*` rules with `languageOptions.nesting: 'standard'`

### ISSUE THREAD

**Issue by romainmenke** [stylelint/stylelint#8906]
While working on https://github.com/stylelint/stylelint/pull/8858 I got a bit confused around the `selector-max-*` rules.

These all seem to count that any selector does not exceed the allowed number of some selector part.

(I personally do not use these, so I am unsure what the use case is)

These rules consider functional pseudo's to be a different context.
This part really confuses me, because that is simply incorrect from any point of view. The whole selector works together. Wrapping anything in a functional pseudo does not make it not part of that selector.

Is this maybe purely because `:not()` was the first such functional pseudo? And people considered `.foo:not(#bar)` as a selector that does not select an ID, so it shouldn't count towards `selector-max-id`?

I am not suggesting that we remove this, if only because this is behavior that has shipped for a long time and many people will be useing it.

But because it goes against the very nature of selectors, it gives unexpected outcomes when we do follow the CSS specification for seemingly unrelated aspects. For example when using standard nesting.

-----

With `languageOptions.nesting: 'standard'` any sufficiently complicated nested selector will resolve with `:is()` (e.g. `a { b { c {} } }` is `:is(a b) c {}`).

That means that all the `selector-max-*` rules suddenly give very surprising results when using standard nesting as most of the selector will be wrapped in `:is()`.

That the selector is wrapped in `:is()` is invisible to users.

-----

This is further complicated by the fact that these rules have rule options to completely ignore functional pseudo's.

-----

I am unsure how to proceed here.

The most viable option seems to leave those rules as-is and keep using `flattenNestedSelectorsForRule` for `selector-max-type`, `selector-max-id`, `selector-max-class`, ...

This seems to be most in the spirit of those rules, to naively combine them and count the number of parts on "the top level".

The only rule (as far as I can tell) where this would produce incorrect results is `selector-max-type`

```css
a {
  b& {}
}
```

Is considered 1 type with `flattenNestedSelectorsForRule`, while it actually is `2`.
But I don't think this is a problem in practice. `a:is(b) {}` doesn't match anything anyway.

Users could be warned about such problematic selectors with a different rule.

**Comment by romainmenke** [comment]
Also important to consider that we can't easily add support for `languageOptions.nesting` in new rules in a patch/minor update.

Users that set `languageOptions.nesting: 'standard'` while also depending on the `nonstandard` behavior for `selector-max-id` would experience a breaking change if we later do add support for it in that rule.

So maybe we should try to cover all rules and find a way to make them work more intuitively with standard nesting? Although I really do not see a way to combine standard nesting with `:is()` and that these rules consider functional pseudo's a different context.

**Comment by jeddy3** [comment]
Thank you for continuing to work on implementing standard nesting and for writing up this issue so clearly! 

> Is this maybe purely because `:not()` was the first such functional pseudo? 

Yes, that's my recollection. The rules predate other functional pseudos like `:is()`, `:where()` and `:has()`, and we took a naive and misjudged approach to `:not()`.

>  I am unsure what the use case is

If I recall, it started with `selector-no-id`, which became `selector-max-id`, as id selectors were considered problematic because of their high specificity and, before the likes of `:where()` and `:is()`, no way to mitigate that. We then added the other `selector-max-*` rules to allow people to:

- limit selector complexity, e.g. limiting the specificity, nesting depth and number of selectors & combinations
- enforce methodologies, e.g. disallowing everything but single class selectors in CSS Modules

We mistakenly added some of the rules for completionism and are seldom used, e.g. `selector-max-pseudo-element`.

`"selector-max-id": 0` looks to be the rule and value that people use most often.

> So maybe we should try to cover all rules and find a way to make them work more intuitively with standard nesting? Although I really do not see a way to combine standard nesting with `:is()` and that these rules consider functional pseudo's a different context.

We have an opportunity to bring these up to date in this major release. It's been a couple of years since our last major release, and the upcoming one is shaping up to be less disruptive than we initially anticipated, thanks to `require(ESM)`, which gives us space to introduce rule changes.

The timing feels appropriate, too, as CSS nesting is Baseline 2023, and we've started adding rules related to modern nesting, such as [`nesting-selector-no-missing-scoping-root`](https://stylelint.io/user-guide/rules/nesting-selector-no-missing-scoping-root).

> That means that all the `selector-max-*` rules suddenly give very surprising results when using standard nesting as most of the selector will be wrapped in `:is()`. That the selector is wrapped in `:is()` is invisible to users.

This is a good point. One that made me realise that the current behaviour of the rules feels off, too.

Given:

```css
#foo {
	& a {}
}
```

And:

```js
export default {
	rules: {
		"selector-max-id": 0
	}
};
```

We get:

```shell
1:1-5 error Expected "#foo" to have no more than 0 ID selectors
2:2-5 error Expected "& a" to have no more than 0 ID selectors
```

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIDEAzCEAEwA6A7bALgGQYCGWAvnhSADQgoCWANjAHKkC2ciMAHlwAOrAHRgAzuLrgIuJgHMEIfoIgAnAhlgpSAV2aac+Amv0xx8LHkI2Q4mKzAF1AWk6k+LxlGwIMABmsCKlwKAG48aUg5RnkAMXV3AiVxAgBPVmZGXAIxWQURTgArKXpYQSlEIwwMX1SMh2yCX0tfZlICc2a6PBq69Myml2iFF1TSXChSNR8-No6u32oQCiA).

The second error is unintuitive as the quoted and highlighted selector doesn't contain any ID selectors.

I think it'd be more intuitive and in the spirit of these rules if they checked what the author sees in their source by default, not what the browser (or previously a post-processor) will desugar it into. The only exceptions are the `*-specificity` rules because the (`&`) nesting selector has a bearing on the specificity.

We can then add a secondary option, e.g. `desugarNestedSelectors`, to them to allow people to opt-in to the old behaviour on a rule-by-rule basis. We can then clearly document the caveats of doing so under the option in each README, e.g. that nested selectors are often desugared using `:is()` and that the problem messages will become obtuse.  

So, in summary, we could change the `selector-max-*` (except `selector-max-specificity`) rules so that they:

- check unnested selectors by default
- have `resolveNestedSelectors` secondary options
- consider functional pseudos in the same context

And fix the `*-specificity` rules so they reflect the standards.

Would these changes make the rules correct and more intuitive?

---

In light of the complexity that @romainmenke has uncovered, I'm wondering if we also take a cleaner break with how we cater to language extensions that use nonstandard nesting.

We can support and focus on standard nesting in Stylelint, and provide a [point of extension](https://stylelint.io/about/vision#extensible) for the community to use their own nesting resolver. It'd then be their responsibility to work with the idiosyncrasies of using nonstandard nesting. Similar to the recent proposal to extend csstree's parser (https://github.com/stylelint/stylelint/issues/8718#issuecomment-3182846712) for language extensions.

If we adopt the `selector-max-*`proposal above, we can document that the `desugarNestedSelectors` rule options support the `customNestingResolver` language option, but we can't guarantee its compatibility (which is what we do for other nonstandard things).

```js
export default {
  "languageOptions": {
    "customNestingResolver": ["@stylelint-scss/nesting-resolver"]
  },
  "rules": {
    "selector-max-id": [1, { "desugarNestingSelector": true }]
  }
}
```

**Comment by ybiquitous** [comment]
@romainmenke @jeddy3 Thank you for raising this discussion. I can now have a better understanding of nested selectors and the related rules. Users could have different requirements for nesting, depending on the rules.

In my understanding, people probably would hope:

- `selector-max-*` rules on *syntax* should not consider nesting, such as `selector-max-id` , `selector-max-class`, etc.
  - E.g., with the `selector-max-id` rule, `#foo { & a {} }` should have 1 problem on `#foo`, none on `& a`.
- `selector-max-*` rules on *semantics* should consider nesting, such as `selector-max-specificity`.

> So, in summary, we could change the `selector-max-*` (except `selector-max-specificity`) rules so that they:
> 
> * check unnested selectors by default
> * have `resolveNestedSelectors` secondary options
> * consider functional pseudos in the same context

Basically, the suggestion by @jeddy3 above (https://github.com/stylelint/stylelint/issues/8906#issuecomment-3622332055) sounds reasonable to me, but I'd suggest the additional things, based on my understanding above.

- Support **only** unnested selectors with `selector-max-*` rules on *syntax*.
- Provide the new `resolveNestedSelectors` options only for backward compatibility (with warnings) and remove the options in future major versions (_transitional_ options).

I believe that avoiding too many options could benefit both users and maintainers.

**Comment by jeddy3** [comment]
> and remove the options in future major versions

I'm beginning to think we should make a clean break from resolving nesting and supporting non-standard nesting in `17.0.0`. I suspect that temporarily adding options could confuse users and add complexity for us as maintainers. Especially when those temporary options would introduce the desugaring to `:is()` behaviour that may surprise users.

I've drafted a migration guide entry to help capture a new proposal:

---

## Changed `selector` and `specificity` rules to account for standard CSS nesting and modern functional pseudo-classes

We've changed the `selector` and `specificity` rules to account for CSS nesting and modern functional pseudo-classes, such as `:is()`, `:where()` and `:has()`.

We've changed the following rules:

- `no-duplicate-selector`
- `selector-max-attribute`
- `selector-max-class`
- `selector-max-combinators`
- `selector-max-compound-selectors`
- `selector-max-id`
- `selector-max-pseudo-class`
- `selector-max-type`
- `selector-max-universal`
- `selector-no-qualifying-type`

So that they no longer:

- desugar nesting
- evaluate functional pseudo-classes separately, such as `:is()` and `:not()`

We've changed the `selector-max-specificity` and `no-descending-specificity` rules to adhere to [CSS Nesting specification](https://drafts.csswg.org/css-nesting/#nest-selector):

> The [specificity](https://drafts.csswg.org/selectors-4/#specificity) of the [nesting selector](https://drafts.csswg.org/css-nesting/#nesting-selector) is equal to the largest specificity among the complex selectors in the parent style rule’s selector list (identical to the behavior of [:is()](https://drafts.csswg.org/selectors-4/#matches-pseudo)), or zero if no such selector list exists.

We've removed two secondary options that conflict with the changes above:

- `checkContextFunctionalPseudoClasses` for `selector-max-id` (new default behaviour)
- `resolveNestedSelectors` for `selector-class-pattern` (incompatible with standard nesting)

You should turn the `specificity` rules off in your config if you use a language extension that uses nonstandard nesting, such as SCSS or Less. 

[Standard nesting](https://drafts.csswg.org/css-nesting/#nest-selector) often desugars to `:is()`. For example, the following:

```css
#foo,
.bar {
  a {}
}
```

Is desugared to:

```css
:is(#foo, .bar) a {}
```

This behaviour, when hidden, can lead to surprising lint problems being flagged. By longer desugaring nesting, the selectors you see in your source files are now linted as-is, and we remove any surprises.

---

Additionally, we can add "This rule is only appropriate for CSS. You should not turn it on for CSS-like languages, such as SCSS or Less." to the READMEs of the `selector-max-specificity` and `no-descending-specificity` rules, as we do for a few other other rules like [media-query-no-invalid](https://stylelint.io/user-guide/rules/media-query-no-invalid/).

The SCSS community can, as they already do for other rules, create their own version of these rules (and `selector-class-pattern`) that would use nonstandard nesting.

This means we can avoid the complexity and potential confusion of introducing a `nesting` `languageOptions` property. We originally intended to include the option because of the number of rules resolving nesting behind the scenes. As this proposal takes out resolving nesting from most rules, there's less need for it.

**Comment by ybiquitous** [comment]
@jeddy3 Thank you for the proposal. Your suggestion sounds more reasonable than mine, especially for the long-term perspective. 👍🏼 

I'm still a bit worried about how many `stylelint-scss` users the clean breaking change would affect. To mitigate such impacts of Stylelint 17.0.0, we probably should support many `stylelint-scss` rules to handle nesting selectors. Fortunately, we have kept a good relationship with maintainers in [stylelint-scss/stylelint-scss](https://github.com/stylelint-scss/stylelint-scss). Anyway, I believe it would be a necessary cost to maintain this project in the long term.

Additionally, since your proposed description of the nesting is awesome, it might be great if we could add a dedicated page in the user guide to explain how nesting selectors work with the built-in Stylelint rules. Such a page could be referred to not only by the v17 migration guide, but also by each rule page handling this nesting.

**Comment by jeddy3** [comment]
I've opened a WIP PR for the proposal: https://github.com/stylelint/stylelint/pull/8913

If we've happy to proceed, I'll open an issue over at `stylelint-scss` to let the know that they may want to create their own versions of `selector-max-specificity`, `no-descending-specificity` and `selector-class-pattern` as these are the only rules that will have a different behaviour for SCSS, or at least turn them off in their shared config that most SCSS authors extend.

**Comment by romainmenke** [comment]
Thank you both for working on this @ybiquitous and @jeddy3  🙇 

I really like the distinction between rules for semantics or for syntax.
